### PR TITLE
Implement Waypoints

### PR DESCRIPTION
**Pull request by CortexPE**
image

### CODE REVIEW

File: Horion/Command/Commands/WaypointCommand.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	assertTrue(player != nullptr);
+	assertTrue(args->size() > 2);
+
+	static auto mod = moduleMgr->getModule<Waypoints>();
+	if (mod == nullptr)
+		return true;
+
+	auto opt = args->at(1);
+	std::string name = args->at(2);
+	assertTrue(name.length() > 0);
```
Comment (richardletshacks) [inline Horion/Command/Commands/WaypointCommand.cpp]: I don't think this right here is needed

File: Horion/DrawUtils.cpp [hunk 2]
```diff
@@ -506,3 +506,25 @@ void DrawUtils::drawLinestrip3d(const std::vector<vec3_t>& points) {
 
 	tess_end(game3dContext, myTess, entityFlatStaticMaterial);
 }
+
+void DrawUtils::drawHologram(vec3_t pos, std::string text, float size) {
```
Comment (richardletshacks) [inline Horion/DrawUtils.cpp:510]: Creating a new function in DrawUtils thats basically just `drawNameTags()` (which already was borderline) with changed colors is really not needed

**Review by richardletshacks — COMMENTED** [review]
good addition, I approve

File: Horion/DrawUtils.cpp (hunk 2 above)
Comment (CortexPE) [reply Horion/DrawUtils.cpp:510]: There isn't a way to draw a hologram/nametag without the entity... which is why I had to add `DrawUtils::drawHologram()`

And using black and 0.5 for transparency would look better for that purpose

**Comment by GodGamer029** [comment]
I like the addition, but adding a new method to DrawUtils is overkill

**Comment by ghost** [comment]
You should add the option to specify what coords you want your waypoint to be, currently you can only set waypoints to your current position

**Comment by CortexPE** [comment]
> You should add the option to specify what coords you want your waypoint to be, currently you can only set waypoints to your current position

you got it

**Comment by CortexPE** [comment]
> I like the addition, but adding a new method to DrawUtils is overkill

in that case, it's appropriate to refactor nametag drawing and "hologram" drawing on DrawUtils so that it doesn't have to use copy-paste code from the same file...

and about nametags... we might want to actually hide the actual vanilla nametags themselves so that it doesn't look like a total mess when we just overlay the horion nametags on top

**Comment by richardletshacks** [comment]
> and about nametags... we might want to actually hide the actual vanilla nametags themselves so that it doesn't look like a total mess when we just overlay the horion nametags on top

it should already do that

**Comment by CortexPE** [comment]
it apparently doesn't... oof

nvm, just for floating text particles (invisible 0.001 scaled client-side human entities)

File: Horion/Command/Commands/WaypointCommand.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+bool WaypointCommand::execute(std::vector<std::string>* args) {
+	C_LocalPlayer* player = g_Data.getLocalPlayer();
+	assertTrue(player != nullptr);
+	assertTrue(args->size() > 2);
+
+	static auto mod = moduleMgr->getModule<Waypoints>();
+	if (mod == nullptr)
+		return true;
```
Comment (GodGamer029) [inline Horion/Command/Commands/WaypointCommand.cpp]: unnecessary check

File: Horion/Command/Commands/WaypointCommand.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	std::string opt = args->at(1);
+	std::string name = args->at(2);
+
+	if (opt == "add") {
+		vec3_t pos = player->currentPos;
+		if (args->size() == 6) {
+			pos.x = assertInt(args->at(3));
+			pos.y = assertInt(args->at(4));
+			pos.z = assertInt(args->at(5));
```
Comment (GodGamer029) [inline Horion/Command/Commands/WaypointCommand.cpp]: What about floats?

File: Horion/Module/Modules/Compass.cpp [hunk 5]
```diff
@@ -17,8 +20,24 @@ void Compass::onPreRender(C_MinecraftUIRenderContext* renderCtx) {
 
 	if (player == nullptr || !GameData::canUseMoveKeys()) return;
 
+	static auto wpMod = moduleMgr->getModule<Waypoints>();
+
+	auto extraPoints = std::multimap<int, std::string>{};
+
+	if (wpMod != nullptr && showWaypoints) {
```
Comment (GodGamer029) [inline Horion/Module/Modules/Compass.cpp]: `wpMod != nullptr` is unnecessary

File: Horion/Module/Modules/Compass.cpp [hunk 6]
```diff
@@ -65,6 +84,29 @@ void Compass::onPreRender(C_MinecraftUIRenderContext* renderCtx) {
 			drawCenteredText(vec2_t(xOff, 30), "NW", 1, majorOpacity);
 			break;
 		}
+		typedef std::multimap<int, std::string>::iterator multimap_iter;
+		std::pair<multimap_iter, multimap_iter> result = extraPoints.equal_range(oDeg);
```
Comment (GodGamer029) [inline Horion/Module/Modules/Compass.cpp]: should probably have used `auto` here

File: Horion/Module/Modules/Waypoints.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		for (const auto& _wp : waypoints) {
+			if (name == _wp.first) {
+				waypoints.erase(name);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	vec3_t* getWaypoint(std::string name) {
```
Comment (GodGamer029) [inline Horion/Module/Modules/Waypoints.h]: should return `vec3_t`, not `vec3_t*`

Comment (GodGamer029) [reply Horion/Module/Modules/Waypoints.h]: you could also use std::optional

File: Horion/Module/Modules/Waypoints.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		if (waypoints.find(name) == waypoints.end())
+			return nullptr;
+
+		return &waypoints[name];
+	};
+
+	std::map<std::string, vec3_t>* getWaypoints() {
+		return &waypoints;
```
Comment (GodGamer029) [inline Horion/Module/Modules/Waypoints.h]: either return the value, or make `waypoints` a std::shared_ptr and return that

**Review by GodGamer029 — CHANGES_REQUESTED** [review]
Imo we can merge already, but i dislike the method in DrawUtils, the method will only be used once and can easily be implemented where its actually used

File: Horion/DrawUtils.cpp (hunk 2 above)
Comment (BootifulApple) [reply Horion/DrawUtils.cpp:510]: nice! any way to add like color to the waypoint or maybe edit it in a gui?

**Pull request merged**